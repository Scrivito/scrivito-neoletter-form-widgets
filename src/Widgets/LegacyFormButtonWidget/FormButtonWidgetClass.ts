import * as Scrivito from "scrivito";

export const FormButtonWidget = Scrivito.provideWidgetClass(
  "FormButtonWidget",
  {
    attributes: {
      buttonText: "string",
      alignment: ["enum", { values: ["left", "center", "right", "block"] }],
    },
    extractTextAttributes: ["buttonText"],
  }
);
