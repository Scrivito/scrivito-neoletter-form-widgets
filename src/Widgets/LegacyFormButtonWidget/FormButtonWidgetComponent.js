import * as React from "react";
import * as Scrivito from "scrivito";
Scrivito.provideComponent("FormButtonWidget", ({ widget }) => (
  <WrapIfClassName className={alignmentClassName(widget.get("alignment"))}>
    <button
      className={`btn btn-primary${
        widget.get("alignment") === "block" ? " btn-block" : ""
      }`}
      type="submit"
    >
      <Scrivito.ContentTag tag="span" content={widget} attribute="buttonText" />
    </button>
  </WrapIfClassName>
));

function WrapIfClassName({ className, children }) {
  return className ? <div className={className}>{children}</div> : children;
}
function alignmentClassName(widgetAlignment) {
  if (widgetAlignment === "center") return "text-center";
  if (widgetAlignment === "right") return "text-end";

  return null;
}
