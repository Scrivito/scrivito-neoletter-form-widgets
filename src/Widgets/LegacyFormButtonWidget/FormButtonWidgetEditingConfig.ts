import * as Scrivito from "scrivito";

Scrivito.provideEditingConfig("FormButtonWidget", {
  title: "Legacy Form Button",
  hideInSelectionDialogs: true,
  attributes: {
    buttonText: {
      title: "Button Text"
    },
    alignment: {
      title: "Alignment",
      values: [
        { value: "left", title: "Left" },
        { value: "center", title: "Center" },
        { value: "right", title: "Right" },
        { value: "block", title: "Full width" }
      ]
    }
  },
  properties: ["buttonText", "alignment"],
  initialContent: {
    buttonText: "send message",
    alignment: "center"
  },
  validations: [
    [
      "alignment",
      alignment => {
        if (!alignment) {
          return "Select the alignment.";
        }
      }
    ]
  ]
});
