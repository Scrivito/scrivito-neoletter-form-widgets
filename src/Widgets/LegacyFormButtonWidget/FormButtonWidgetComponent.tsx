import * as React from "react";
import * as Scrivito from "scrivito";
import { FormButtonWidget } from "./FormButtonWidgetClass";
interface WrapIfClassNameProps {
  className: string;
}

Scrivito.provideComponent(FormButtonWidget, ({ widget }) => (
  <WrapIfClassName className={alignmentClassName(widget.get("alignment"))}>
    <button
      className={`btn btn-primary${
        widget.get("alignment") === "block" ? " btn-block" : ""
      }`}
      type="submit">
      <Scrivito.ContentTag tag="span" content={widget} attribute="buttonText" />
    </button>
  </WrapIfClassName>
));

function WrapIfClassName({
  className,
  children
}: React.PropsWithChildren<WrapIfClassNameProps>) {
  return className ? <div className={className}>{children}</div> : children;
}
function alignmentClassName(widgetAlignment: string | null) {
  if (widgetAlignment === "center") return "text-center";
  if (widgetAlignment === "right") return "text-end";
  return "";
}
