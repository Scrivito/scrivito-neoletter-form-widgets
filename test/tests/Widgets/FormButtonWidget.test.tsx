import * as Scrivito from "scrivito";
import { screen } from "@testing-library/react";
import { FormButtonWidget } from "../../../src/Widgets/LegacyFormButtonWidget/FormButtonWidgetClass";
import PageRenderer from "../../helpers/pageRenderer";
import "../../../src/Widgets/LegacyFormButtonWidget/FormButtonWidgetComponent";

Scrivito.configure({ tenant: "inMemory" });

const pageRenderer = new PageRenderer();

describe("FormButtonWidget", () => {
  it("renders button with default alignment", () => {
    const widgetProps = {
      alignment: null,
      buttonText: "Submit"
    };

    pageRenderer.render({
      body: [new FormButtonWidget(widgetProps)]
    });

    const buttonElement = screen.getByRole("button");
    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toHaveTextContent("Submit");
  });

  it("renders button with center alignment", () => {
    const widgetProps = {
      alignment: "center",
      buttonText: "Center Button"
    };

    pageRenderer.render({
      body: [new FormButtonWidget(widgetProps)]
    });

    const buttonElement = screen.getByRole("button");
    const container = document.querySelector(".text-center");
    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toHaveTextContent("Center Button");
    expect(container).toHaveClass("text-center");
  });

  it("renders button with block alignment", () => {
    const widgetProps = {
      alignment: "block",
      buttonText: "Block Button"
    };

    pageRenderer.render({
      body: [new FormButtonWidget(widgetProps)]
    });

    const buttonElement = screen.getByRole("button");
    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toHaveTextContent("Block Button");
    expect(buttonElement.className).toContain("btn-block");
  });

  it("renders button with right alignment", () => {
    const widgetProps = {
      alignment: "right",
      buttonText: "Right Button"
    };

    pageRenderer.render({
      body: [new FormButtonWidget(widgetProps)]
    });

    const buttonElement = screen.getByRole("button");
    const container = document.querySelector(".text-end");
    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toHaveTextContent("Right Button");
    expect(container).toHaveClass("text-end");
  });

  it("renders correctly", () => {
    const widgetProps = {
      alignment: "center",
      buttonText: "Center Button"
    };
    const tree = pageRenderer.getAsJSON({
      body: [new FormButtonWidget(widgetProps)]
    });
    expect(tree).toMatchSnapshot();
  });
});
