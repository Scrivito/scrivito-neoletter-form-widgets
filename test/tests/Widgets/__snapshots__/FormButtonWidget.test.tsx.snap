// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormButtonWidget renders correctly 1`] = `
<div>
  <div>
    <div
      className="text-center"
    >
      <button
        className="btn btn-primary"
        type="submit"
      >
        <span>
          Center Button
        </span>
      </button>
    </div>
  </div>
</div>
`;
